src/connector/src/sink/kafka.rs

```diff
@@ -0,0 +1,557 @@
+        let key = field.name.clone();
+        let value = datum_to_json_object(field, datum_ref)
+            .map_err(|e| SinkError::JsonParse(e.to_string()))?;
+        mappings.insert(key, value);
+    }
+    Ok(mappings)
+}
+
+pub fn chunk_to_json(chunk: StreamChunk, schema: &Schema) -> Result<Vec<String>> {
+    let mut records: Vec<String> = Vec::with_capacity(chunk.capacity());
+    for (_, row) in chunk.rows() {
+        let record = Value::Object(record_to_json(row, schema.fields.clone())?);
+        records.push(record.to_string());
+    }
+
+    Ok(records)
+}
+
+/// the struct conducts all transactions with Kafka
+#[derive(Clone)]
+pub struct KafkaTransactionConductor {
+    pub properties: KafkaConfig,
+    pub inner: Arc<ThreadedProducer<DefaultProducerContext>>,
+    in_transaction: bool,
+}
+
+impl KafkaTransactionConductor {
+    async fn new(config: KafkaConfig) -> Result<Self> {
+        let inner = ClientConfig::new()
+            .set("bootstrap.servers", config.brokers.as_str())
+            .set("message.timeout.ms", "5000")
+            .set("transactional.id", config.identifier.as_str()) // required by kafka transaction
+            .create_with_context(DefaultProducerContext)
+            .expect("Producer creation error");
+
+        Ok(KafkaTransactionConductor {
+            properties: config,
+            inner: Arc::new(inner),
+            in_transaction: false,
+        })
+    }
+
+    fn init_transaction(&self) -> impl Future<Output = KafkaResult<()>> {
+        let inner = self.inner.clone();
+        let timeout = self.properties.timeout;
+        task::spawn_blocking(move || inner.init_transactions(timeout))
+            .unwrap_or_else(|_| Err(KafkaError::Canceled))
+    }
+
+    fn start_transaction(&self) -> impl Future<Output = KafkaResult<()>> {
+        let inner = Arc::clone(&self.inner);
+        task::spawn_blocking(move || inner.begin_transaction())
+            .unwrap_or_else(|_| Err(KafkaError::Canceled))
+    }
+
+    fn commit_transaction(&self) -> impl Future<Output = KafkaResult<()>> {
+        let inner = Arc::clone(&self.inner);
+        let timeout = self.properties.timeout;
+        task::spawn_blocking(move || inner.commit_transaction(timeout))
+            .unwrap_or_else(|_| Err(KafkaError::Canceled))
+    }
+
+    fn abort_transaction(&self) -> impl Future<Output = KafkaResult<()>> {
+        let inner = Arc::clone(&self.inner);
+        let timeout = self.properties.timeout;
+        task::spawn_blocking(move || inner.abort_transaction(timeout))
+            .unwrap_or_else(|_| Err(KafkaError::Canceled))
+    }
+
+    fn flush(&self) -> impl Future<Output = KafkaResult<()>> {
+        let inner = Arc::clone(&self.inner);
+        let timeout = self.properties.timeout;
+        task::spawn_blocking(move || inner.flush(timeout))
+            .map_ok(|_| KafkaResult::Ok(()))
+            .unwrap_or_else(|_| Err(KafkaError::Canceled))
```

Comment:
Is there any way to prevent spawn blocking?

Reply:
why? I think we can reuse more code this way